### PR TITLE
Fix Console.Write on iOS

### LINKED ISSUE #36440: Console.Write prints a new line on iOS
Example:

```csharp
Console.Write ("A");
Console.Write ("B");
Console.WriteLine ("C");
```

Output with .NET 5:

```
2020-05-14 12:50:56.386845+0200 MySingleView[89908:5086139] A
2020-05-14 12:50:56.386961+0200 MySingleView[89908:5086139] B
2020-05-14 12:50:56.387585+0200 MySingleView[89908:5086139] C
```

Output with current Xamarin.iOS:

```
2020-05-14 12:53:03.497594+0200 testapp[93230:5096504] ABC
```

The problem seems to be that mono's implementation stores text until a newline is written:

https://github.com/mono/mono/blob/f24a85069b3541ef45c8bbbd800ec7d88ad22e07/mcs/class/corlib/System/Console.iOS.cs#L81-L83

while dotnet's implementation writes out the text every time:

https://github.com/dotnet/runtime/blob/622341ecc2d43ef81c509bde53de8513727d626d/src/libraries/System.Console/src/System/ConsolePal.iOS.cs#L16-L24

and NSLog will append a newline if the text doesn't have one.

### PR DESCRIPTION
**Pull request by MaximLipnin**
Fixes https://github.com/dotnet/runtime/issues/36440.

This is iOS-part of Egor's fix (with Ralf's suggestion) from https://github.com/dotnet/runtime/pull/37078.

For this code snippet:
```
Console.Write ("A");
Console.Write ("B");
Console.WriteLine ("C");

Console.Write ("D\nd");
Console.Write ("E\ne");
Console.WriteLine ("F");
```
the output will be:
```
2021-08-02 12:19:25.865365+0300 iOS.Simulator.ConsoleWrite.Test[4681:132934] ABC
2021-08-02 12:19:25.868503+0300 iOS.Simulator.ConsoleWrite.Test[4681:132934] D
2021-08-02 12:19:25.869020+0300 iOS.Simulator.ConsoleWrite.Test[4681:132934] dE
2021-08-02 12:19:25.869138+0300 iOS.Simulator.ConsoleWrite.Test[4681:132934] eF
```

### CODE REVIEW

File: src/libraries/System.Console/src/System/IO/ConsoleStream.cs [hunk 1]
```diff
@@ -90,5 +91,37 @@ protected void ValidateWrite(byte[] buffer, int offset, int count)
                 throw Error.GetWriteNotSupported();
             }
         }
+
+        protected static void AccumulateNewLines(StringBuilder accumulator, ReadOnlySpan<char> buffer, Action<string> printer)
```
Comment (adamsitnik) [inline src/libraries/System.Console/src/System/IO/ConsoleStream.cs:95]: if this method is specifc to iOS and not reused anywhere else, why not to keep it in `ConsolePal.iOS.cs` and replace `Action<string> printer` with direct calls to `Interop.Sys.Log`?

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 2]
```diff
@@ -1,22 +1,34 @@
...
+            ReadOnlySpan<char> charSpan = MemoryMarshal.Cast<byte, char>(buffer);
+
+            lock (_buffer)
             {
-                Interop.Sys.Log(ptr, buffer.Length);
+                AccumulateNewLines(_buffer, charSpan, line =>
+                {
+                    fixed (char* ptr = line)
+                    {
+                        Interop.Sys.Log((byte*)ptr, line.Length * 2);
```
Comment (adamsitnik) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:29]: Is it possible to somehow force `NSLog` to stop appending new line for every logged message?

**Comment by am11** [comment]
Can we enable https://github.com/dotnet/runtime/blob/79c4144f174754bf4fa18827a292549214f86581/src/libraries/System.Console/tests/ConsoleEncoding.cs#L25-L26
or any other test for this on iOS? (from the skip reason message of this test, it's not clear which API is not supported)

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 2 above)
Comment (steveisok) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:29]: Not that I know of.  The only options on `NSLog` are string formatting ones.

File: src/libraries/System.Console/src/System/IO/ConsoleStream.cs (hunk 1 above)
Comment (akoeplinger) [reply src/libraries/System.Console/src/System/IO/ConsoleStream.cs:95]: Yeah I think we should move it to ConsolePal.iOS.cs since we no longer need it on Android.

We should also implement Rolf's suggestion from https://github.com/dotnet/runtime/pull/37078#discussion_r431814922 to walk the buffer backwards.

File: src/libraries/System.Console/src/System/IO/ConsoleStream.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 using System.Diagnostics;
 using System.Runtime.InteropServices;
+using System.Text;
```
Comment (am11) [inline src/libraries/System.Console/src/System/IO/ConsoleStream.cs:6]: nit: unused now
```suggestion
```

Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/IO/ConsoleStream.cs]: fixed, thanks

**Comment by MaximLipnin** [comment]
> Can we enable
> 
> https://github.com/dotnet/runtime/blob/79c4144f174754bf4fa18827a292549214f86581/src/libraries/System.Console/tests/ConsoleEncoding.cs#L25-L26
> 
> 
> or any other test for this on iOS? (from the skip reason message of this test, it's not clear which API is not supported)

Added `TestConsoleWrite` test

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 4]
```diff
@@ -1,22 +1,66 @@
...
+        private readonly StringBuilder _buffer = new StringBuilder();
+
         public NSLogStream() : base(FileAccess.Write) {}
 
         public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
 
         public override unsafe void Write(ReadOnlySpan<byte> buffer)
         {
-            fixed (byte* ptr = buffer)
+            ReadOnlySpan<char> charSpan = MemoryMarshal.Cast<byte, char>(buffer);
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: This doesn't look valid.  Why is it guaranteed that the input bytes are actually two-byte chars?

Comment (akoeplinger) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: if I understand things correctly this is because we set `OutputEncoding => Encoding.Unicode;` on the `ConsolePal`.

Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: Then what happens when someone does:
```C#
Stream s = Console.OpenOutputStream();
byte[] bytes = Encoding.Unicode.GetBytes("s");
s.WriteByte(byte[0]);
s.Writebyte(byte[1]);
```
?

Comment (akoeplinger) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: Yeah that wouldn't work...
I guess we could override `WriteByte()` and throw an exception?

Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: It's not just WriteByte. It's `Write(...)` that doesn't write out on the exact alignment expected.  Seems very fragile.

Comment (akoeplinger) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: Right, do you think adding a check for `buffer.Length % 2 == 0` would be enough?

Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: And then do what if it's not?  Are you suggesting it throw even though it's technically valid use?

Comment (akoeplinger) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: Yes, I'd throw an exception, not sure what else we could do?

Comment (am11) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: (perhaps not optimal for correctness, but) instead of byte->string->byte, can we keep it as byte and use `MemoryStream` for accumulation and then `printer(stream.ToArray(), index etc. ...)`?

Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: > not sure what else we could do?

Buffer the remaining bytes for use on the next call?  Presumably by having a Decoder as a field in the stream and running all bytes through that.
```C#
using System.Text;

var decoder = Encoding.Unicode.GetDecoder();
char[] chars = new char[100];
Span<byte> bytes = stackalloc byte[1];
foreach (byte b in Encoding.Unicode.GetBytes("This is a test. This is only a test."))
{
    bytes[0] = b;
    int i = decoder.GetChars(bytes, chars, flush: false);
    if (i > 0)
    {
        Console.WriteLine($"Got {i}: {new string(chars, 0, i)}");
    }
}
```

Comment (adamsitnik) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:21]: `NSLogStream` is used to create `StreamWriter` in `Console.CreateOutputWriter`:

https://github.com/dotnet/runtime/blob/f1bf5eee390fa118441cae14a157773f1e2d751d/src/libraries/System.Console/src/System/Console.cs#L240-L250

`StreamWriter` just appends all characters to a private buffer and encodes them to bytes on every `Flush` and in case of `Console` it's for every write (autoflush==true):

https://github.com/dotnet/runtime/blob/f1bf5eee390fa118441cae14a157773f1e2d751d/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L485-L487

perhaps instead of modyfing `NSLogStream` we could create a custom type that derives from `StreamWriter` and change its flushing logic (by overriding `Flush`) to do similar thing that @MaximLipnin wrote here:
* take the private buffer
* find next line
* encode it to bytes
* call Write(bytes)
* repeat until the buffer is empty

**Comment by MaximLipnin** [comment]
It's quite close to a working state, the only problem is that I cannot find the right moment to write the data remaining in the cache to the stream before closing the app.

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs [hunk 5]
```diff
@@ -320,11 +323,55 @@ private void Flush(bool flushStream, bool flushEncoder)
...
+                    if (count > 0)
+                    {
+                        _acc.AddRange(byteBuffer.Slice(0, count).ToArray());
+                    }
+                }
+
+                _charPos = 0;
+                _lineStartPos = 0;
+
+                if (flushStream && flushEncoder)
```
Comment (MaximLipnin) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs:357]: I thought that this is the right moment to write the remaining data, but it's not, e.g. in case :
```
    Console.Write("AB\nCD");
    Console.Write("EF\nGH");
```
`GH` is being missed.

File: src/libraries/System.Console/tests/ReadAndWrite.cs [hunk 6]
```diff
@@ -194,6 +194,25 @@ public static async Task OutWriteAndWriteLineOverloads()
         }
     }
 
+    [Fact]
+    [PlatformSpecific(TestPlatforms.iOS | TestPlatforms.MacCatalyst | TestPlatforms.tvOS)]
+    public void TestConsoleWrite()
```
Comment (mdh1418) [inline src/libraries/System.Console/tests/ReadAndWrite.cs:199]: Would it be better to align this test with the code snippet and example output and incorporate `\n`
```
Console.Write ("A");
Console.Write ("B");
Console.WriteLine ("C");

Console.Write ("D\nd");
Console.Write ("E\ne");
Console.WriteLine ("F");
```

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs [hunk 7]
```diff
@@ -36,6 +37,8 @@ public class StreamWriter : TextWriter
         private readonly char[] _charBuffer;
         private int _charPos;
         private int _charLen;
+        private List<byte> _acc = new List<byte>();
```
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs:40]: nit: `readonly`
```suggestion
        private readonly char[] _charBuffer;
        private readonly List<byte> _acc = new List<byte>();
        private int _charPos;
        private int _charLen;
```

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs [hunk 8]
```diff
@@ -320,11 +323,55 @@ private void Flush(bool flushStream, bool flushEncoder)
...
+                {
+                    if (_charBuffer[i] == '\n')
+                    {
+                        int count = _encoder.GetBytes(new ReadOnlySpan<char>(_charBuffer, _lineStartPos, i - _lineStartPos), byteBuffer, flushEncoder);
+                        if (count > 0)
+                        {
+                            _acc.AddRange(byteBuffer.Slice(0, count).ToArray());
+                        }
+
+                        _stream.Write(_acc.ToArray().AsSpan<byte>());
```
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs:338]: nit: avoid ToArray()
```suggestion
                        _stream.Write(CollectionsMarshal.AsSpan(_acc));
```

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs [hunk 9]
```diff
@@ -320,11 +323,55 @@ private void Flush(bool flushStream, bool flushEncoder)
                     (_byteBuffer = new byte[_encoding.GetMaxByteCount(_charBuffer.Length)]);
             }
 
-            int count = _encoder.GetBytes(new ReadOnlySpan<char>(_charBuffer, 0, _charPos), byteBuffer, flushEncoder);
-            _charPos = 0;
-            if (count > 0)
+            if (OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsMacCatalyst())
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs:326]: `StreamWriter` is a public class that can be used for many things. This particular logic changes it's behaviour for few OSes. Could you please create a custom type that derives from `StreamWriter` ,  override `Flush` and move the new logic there? You might need to make some of its private fields `protected internal`.

https://github.com/dotnet/runtime/blob/4d26112c07391f563cb0ca46af1993635399007e/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L277

**Review by adamsitnik — CHANGES_REQUESTED** [review]

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 10]
```diff
@@ -1,7 +1,9 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
 using System.IO;
+using System.Runtime.InteropServices;
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:6]: Are these changes necessary?

File: src/libraries/System.Console/src/System/IO/StreamWriter.iOS.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+            innerCharPos = 0;
+            _lineStartPos = 0;
+
+            if (flushStream && flushEncoder)
+            {
+                var accArray = _acc.ToArray();
+                if (accArray.Length > 0)
+                {
+                    innerStream.Write(accArray.AsSpan<byte>().Slice(0, accArray.Length));
```
Comment (am11) [inline src/libraries/System.Console/src/System/IO/StreamWriter.iOS.cs:93]: nit:
```suggestion
                var accSpan = CollectionsMarshal.AsSpan(_acc);
                if (accSpan.Length > 0)
                {
                    innerStream.Write(accSpan.Slice(0, accSpan.Length));
```

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs [hunk 12]
```diff
@@ -29,15 +30,15 @@ public class StreamWriter : TextWriter
...
-        private readonly Encoder _encoder;
-        private byte[]? _byteBuffer;
-        private readonly char[] _charBuffer;
-        private int _charPos;
+        protected internal readonly Stream innerStream;
+        protected internal readonly Encoding innerEncoding;
+        protected internal readonly Encoder innerEncoder;
+        protected internal byte[]? innerByteBuffer;
+        protected internal readonly char[] innerCharBuffer;
+        protected internal int innerCharPos;
```
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs:38]: style nit: `protected internal` fields are in PascalCasing elsewhere.

File: src/libraries/System.Console/src/System/IO/StreamWriter.iOS.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.IO
+{
+    internal class IOSStreamWriter : StreamWriter
+    {
+        private readonly List<byte> _acc = new List<byte>();
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/IO/StreamWriter.iOS.cs:12]: What does `acc` mean/stand for? I think we should use whole words for the field.

File: src/libraries/System.Console/src/System.Console.csproj [hunk 14]
```diff
@@ -25,6 +30,7 @@
              Link="Common\Interop\Unix\Interop.Log.cs" />
     <Compile Include="$(CommonPath)Interop\Unix\Interop.Libraries.cs"
              Link="Common\Interop\Unix\Interop.Libraries.cs" />
+    <Compile Include="System\IO\StreamWriter.iOS.cs" />
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System.Console.csproj:33]: Does this file need to be included when `TargetsMacCatalyst=true`?

**Review by eerhardt — DISMISSED** [review]
I don't believe we should be changing the public surface area of StreamWriter in this change. We should find a different way to accomplish this without changing the public surface area.

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 15]
```diff
@@ -1,23 +1,89 @@
...
+            _decoder = Encoding.Unicode.GetDecoder();
+        }
 
         public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
 
         public override unsafe void Write(ReadOnlySpan<byte> buffer)
         {
-            fixed (byte* ptr = buffer)
+            int charCount = _decoder.GetCharCount(buffer, false);
+            Span<char> charSpan = new char[charCount];
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:26]: Can this array be pooled?

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 16]
```diff
@@ -1,23 +1,89 @@
...
+                }
+            }
+
+            if (lineStartIndex < charBuffer.Length)
+            {
+                // no newlines found, add the entire buffer to the cache
+                cache.Append(charBuffer.Slice(lineStartIndex, charBuffer.Length - lineStartIndex));
+            }
+
+            static void printer(string line)
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:62]: Method names should be PascalCased. And should be verbs.

```suggestion
            static void Print(string line)
```

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 17]
```diff
@@ -1,23 +1,89 @@
...
+using System.Runtime.InteropServices;
 using System.Text;
 
 namespace System
 {
     internal sealed class NSLogStream : ConsoleStream
     {
-        public NSLogStream() : base(FileAccess.Write) {}
+        private readonly StringBuilder _buffer = new StringBuilder();
+        private Decoder _decoder;
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:14]:
```suggestion
        private readonly Decoder _decoder;
```

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 18]
```diff
@@ -1,23 +1,89 @@
...
+        {
+            int lineStartIndex = 0;
+            for (int i = 0; i < charBuffer.Length; i++)
+            {
+                if (charBuffer[i] == '\n')
+                {
+                    Span<char> lineSpan = charBuffer.Slice(lineStartIndex, i - lineStartIndex);
+                    if (cache.Length > 0)
+                    {
+                        printer(cache.Append(lineSpan).ToString());
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:44]: I wonder if you can use the `StringBuilder.GetChunks()` method instead of allocating a new string here. You should be able to enumerate through the "chunks" of the StringBuilder, and write them to `Interop.Sys.Log` without allocating.

**Comment by MaximLipnin** [comment]
@eerhardt Thanks for reviewing! I'll address your feedback. There is still a problem of printing the last bits of the cached data before an application exits. I haven't found the proper moment of the code life cycle when it would be possible as `Dispose` doesn't seem to suit this purpose.
The example:
`Console.Write("AB\n");` will print `AB`
but 
`Console.Write("AB");` will print nothing as `AB` will be held in the cache unless a new line appears, which means "forever".

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 18 above)
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:44]: String builder can contain more than 1 chunk so enumerating though and writing them to Interop.Sys.Log means invoking NSLog several times, which would result in the same situation as right now because NSLog automatically appends a new line to any input.

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 15 above)
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:26]: Could you please clarify a bit with an example?

Comment (eerhardt) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:26]: Every time this method is called, it is allocating a new array just to use as a scratch pad, and then thrown away. If we pool the array instead, we don't need to allocate a new array every time.

Here's an example:

https://github.com/dotnet/runtime/blob/60af14837e78d69d5469c2c0c7632cfca761aa09/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L66-L78

Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: Thanks!

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 19]
```diff
@@ -1,22 +1,79 @@
...
+        }
 
         public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
 
         public override unsafe void Write(ReadOnlySpan<byte> buffer)
         {
-            fixed (byte* ptr = buffer)
+            int charCount = _decoder.GetCharCount(buffer, false);
+            char[] pooledBuffer = ArrayPool<char>.Shared.Rent(charCount);
+            Span<char> charSpan = pooledBuffer;
```
Comment (eerhardt) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:28]: One "gotcha" with ArrayPool is that the length of the rented array is guaranteed to be _at least_ the requested length, but it may (and likely will) be larger than what you requested. So it is not a good idea to use the length of the rented array as the same as the requested length.

Below in `WriteOrCache`, it is using `charBuffer.Length`, which will be the length of the rented array.

Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:28]: thanks, I used the length given by the decoder

**Comment by MaximLipnin** [comment]
Trying to override the `Flush` method in the `NSLogStream` class doesn't seem helpful as well because it' being invoked on every Write/WriteLine call due to `AutoFlush` property set to true.

**Comment by MaximLipnin** [comment]
Also, trying to fix the problem from the native side, I gave a try to the `os_log` API (instead of `NSLog`) in a hope it would not append a new line to the output but it did.

**Comment by am11** [comment]
>  `Console.Write("AB");` will print nothing as `AB` will be held in the cache unless a new line appears, which means "forever".

Seems like Xamarin.iOS implementation has the [same limitation](https://github.com/mono/mono/blob/f24a85069b3541ef45c8bbbd800ec7d88ad22e07/mcs/class/corlib/System/Console.iOS.cs#L81-L83), so it won't surprise the existing consumer.

Personally I am not sure if Console.Write*() are the right APIs to achieve the (debugging/tracing -like) characteristics of NSLog in modern .NET. It sounds like a very good candidate of a separate API under `src/libraries/Microsoft.Extensions.Logging*`.

**Review by steveisok — APPROVED** [review]

**Review by adamsitnik — APPROVED** [review]
Overall LGTM, but please check if you could apply my perf suggestion. Thank you @MaximLipnin !

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 20]
```diff
@@ -1,22 +1,85 @@
...
+        private static unsafe void WriteOrCache(StringBuilder cache, Span<char> charBuffer, int length)
+        {
+            for (int i = length - 1; i >= 0; i--)
+            {
+                if (charBuffer[i] == '\n')
+                {
+                    Span<char> lineSpan = charBuffer.Slice(0, i);
+                    if (cache.Length > 0)
+                    {
+                        Print(cache.Append(lineSpan).ToString());
```
[the remark this answers is not in the document]
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: Calling Print on each Span will result in a sequence of the chunk, each of which will be followed by a new line, which is what I'm trying to avoid.

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 21]
```diff
@@ -1,22 +1,83 @@
...
+                    cache.Append(charBuffer.Slice(lastNewLine + 1, length - lastNewLine - 1));
+                }
+
+                return;
+            }
+
+            // no newlines found, add the entire buffer to the cache
+            cache.Append(charBuffer.Slice(0, length));
+
+            static void Print(string line)
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:75]:
```suggestion
            static void Print(ReadOnlySpan<char> line)
```

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 22]
```diff
@@ -1,22 +1,83 @@
...
+            {
+                Span<char> lineSpan = charBuffer.Slice(0, lastNewLine);
+                if (cache.Length > 0)
+                {
+                    Print(cache.Append(lineSpan).ToString());
+                    cache.Clear();
+                }
+                else
+                {
+                    Print(lineSpan.ToString());
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:61]:
```suggestion
                    Print(lineSpan);
```

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: Addressed in the latest commit

**Comment by MaximLipnin** [comment]
The PR looks pretty green now. @stephentoub Would you like to take another look?

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 23]
```diff
@@ -1,22 +1,83 @@
...
-            fixed (byte* ptr = buffer)
+            int maxCharCount = _encoding.GetMaxCharCount(buffer.Length);
+            char[]? pooledBuffer = null;
+            Span<char> charSpan = maxCharCount <= 512 ? stackalloc char[512] : (pooledBuffer = ArrayPool<char>.Shared.Rent(maxCharCount));
+            try
+            {
+                int count = _decoder.GetChars(buffer, charSpan, false);
+                if (count > 0)
+                {
+                    WriteOrCache(_buffer, charSpan, count);
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:36]: This should be changed to slice the char span to the exact used length,  and then not also pass in a count.  Otherwise,  you might be looking at garbage in the helper call,  e.g. with the LastIndexOf.

Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: Addressed, thanks!

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 24]
```diff
@@ -1,22 +1,83 @@
...
+            finally
+            {
+                if (pooledBuffer != null)
+                {
+                    ArrayPool<char>.Shared.Return(pooledBuffer);
+                }
+            }
+        }
+
+        private static unsafe void WriteOrCache(StringBuilder cache, Span<char> charBuffer)
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:48]: Why is the `unsafe` needed?

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 25]
```diff
@@ -1,22 +1,83 @@
...
+                if (lastNewLine + 1 < charBuffer.Length)
+                {
+                    cache.Append(charBuffer.Slice(lastNewLine + 1, charBuffer.Length - lastNewLine - 1));
+                }
+
+                return;
+            }
+
+            // no newlines found, add the entire buffer to the cache
+            cache.Append(charBuffer);
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:73]: General question about this change... it seems like this change means that if I have code that does:
```C#
while (true)
{
    Console.Write(".");
    await Task.Delay(1000);
}
```
to print out a period to the console once a second, for example as you might find with a progress indicator in a console app, that the data will never be emitted to the console, because newlines are never found.  Am I understanding correctly?  Is that not an issue, e.g. because Console is only used for diagnostic messages in an iOS app?

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs [hunk 26]
```diff
@@ -1,22 +1,83 @@
...
+                    cache.Clear();
+                }
+                else
+                {
+                    Print(lineSpan);
+                }
+
+                if (lastNewLine + 1 < charBuffer.Length)
+                {
+                    cache.Append(charBuffer.Slice(lastNewLine + 1, charBuffer.Length - lastNewLine - 1));
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.iOS.cs:66]:
```suggestion
                    cache.Append(charBuffer.Slice(lastNewLine + 1));
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 24 above)
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: yeah, updated, thanks

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 26 above)
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs]: Updated, thanks

File: src/libraries/System.Console/src/System/ConsolePal.iOS.cs (hunk 25 above)
Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/ConsolePal.iOS.cs:73]: Yes, this is a thing I wasn't able to beat. I mentioned it a bit earlier https://github.com/dotnet/runtime/pull/56713#issuecomment-899589419. The current Xamarin implementation has the same drawback so it won't surprise the existing customers.
